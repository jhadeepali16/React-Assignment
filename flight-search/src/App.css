@import url(https://fonts.googleapis.com/css?family=Open+Sans:300,400,700,800);
body{
  margin:0;
  padding:0;
  font-family: Open Sans;
  font-size: 14px;
}
div#root {
  background: transparent url(https://www.airasia.com/cdn/aa-images/main-banner/inen/eatshopfly_repeat_intl.jpg) no-repeat;
  background-size: contain;
  background-position: top center;
  display: flex;
  width: 100%;
  flex-direction: row;
}
main {
  width: 980px;
  margin: 0 auto;
}
.heading h1 {
  margin: 0 auto;
  text-align: left;
  max-width: 980px;
}
.heading {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  background: white;
  z-index: 2;
  box-shadow: 1px 1px 4px rgba(0,0,0,.26);
  padding: 15px 20px;
}
.calendar {
  border: 1px solid;
}
.c-search-engine{
  display: flex;
  justify-content: space-between;
  margin-top: 68px;
}
.c-search-engine .c-flight-search{
  padding: 15px;
  padding-top: 25px;
  width: 40%;
  box-sizing: content-box;
}
.c-search-engine .c-flight-search .element{
  display: flex;
  flex-direction: row;
}
.c-search-engine .c-flight-search .element .tabs {
  background-color: #fff;
  border: 1px solid #e6e6e6;
  border-bottom: 0;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  padding: 10px 30px;
  outline: none;
  cursor: pointer;
  transition: all 0.3s ease-in;
  font-size: 14px;
  font-weight: bold;
}
.c-search-engine .c-flight-search .element .tabs:last-child{
  margin-left: -10px;
}
.c-search-engine .c-flight-search .element .tabs.active {
  background-color: #f1f1f0;
  z-index: 1;
  box-shadow: 2px -3px 6px rgba(0,0,0,.26);
  margin-bottom: -1px;
}
.c-search-engine .c-flight-search .element .tabs:hover {
  background-color: #f1f1f0;
  border-color: rgba(0,0,0,0.1);
}
.c-search-engine .c-flight-search .form{
  display: flex;
  flex-direction: column;
  padding: 10px;
  border: 1px solid #e6e6e6;
  box-shadow: 1px 1px 4px rgba(0,0,0,.26);
  background-color: #f1f1f0;
}
.c-search-engine .c-flight-search .form .element {
  display: flex;
  flex-direction: column;
  text-align: left;
  margin-bottom: 10px;
  position: relative;
}
.form .element label {
  top: 19px;
  position: absolute;
  left: 10px;
  transition: all 0.4s ease-in-out;
  font-size: 14px;
}
.form .element input, .form .element select {
  padding: 18px 10px;
  border: 1px solid #ccc;
  font-family: Open Sans;
  font-size: 14px;
}
.form .element input.error, .form .element select.error{
  border-color:red;
}
.form .element input.error + label, .form .element select.error + label {
  color: red;
}
.form .element span.error {
  color: red;
  font-size: 11px;
}
.form .element input:focus + label, .form .element select:focus + label{
  top: 3px;
  font-size: 12px;
}
.form .element label.active {
  top: 3px;
  font-size: 12px;
}
.c-search-engine .c-flight-search .form .element button {
  border-radius: 0;
  width: auto;
  padding: 10px;
  margin: 0;
  align-self: flex-start;
  border: 1px solid #e6e6e6;
  color:#fff;
  background-color: #00266b;
}
.c-search-engine .c-flight-search .form .element button:hover {
  background-color:#2e77fd;
  outline: 0;
}
.form .element.inactive input{
  pointer-events: none;
  cursor: not-allowed;
  background-color: #f1f1f1;
}
.form .element.inactive label {
  pointer-events: none;
  color: #bdb8b8;
}

.form .element .react-calendar {
  width: 100%;
  transition: all 0.5s ease-in-out;
  position: absolute;
  top: 55px;
  right: 0;
  height: auto;
  z-index: 10;
}
.price-slider {
  padding: 10px 15px 35px;
  margin-top: 20px;
  border: 1px solid #e6e6e6;
  box-shadow: 1px 1px 4px rgba(0,0,0,.26);
  background-color: #f1f1f0;
}
.price-slider h2 {
  text-align: left;
  margin: 5px 0;
}
.price-slider .input-range{
  padding: 10px;
  box-sizing: border-box;
  display: block;
  max-width: 93%;
  margin: 50px auto 0;
}
.input-range__label-container {
  font-family: Open sans;
}
.input-range__track {
  background: #adadad;
}
.input-range__track--active {
  background: #3f51b5;
}

.c-flight-search-result {
  display: flex;
  width: 57%;
  justify-content: center;
  flex-direction: column;
  background: rgba(255, 255, 255, 0.8);
}
.c-flight-search-result p.no-flight{
  color: #000;
  text-align:center;
  font-size: 16px;
  align-self: center;
  font-weight: bold;
}
.c-flight-search-result .trip{
  text-align: left;
  padding: 20px;
  font-size: 18px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.c-flight-search-result .trip .date p{
  margin: 0;
  font-size: 12px;
}
.c-flight-search-result .result .booking{
  display: flex;
  box-shadow: 1px 1px 4px rgba(0,0,0,.26);
  padding: 10px;
  margin-bottom: 20px;
  justify-content: space-between;
  background-color: #fff;
}
.c-flight-search-result .result .booking:hover{
  background-color: #f1f1f0;
}
.c-flight-search-result .result {
  padding:0 20px;
}
.c-flight-search-result .result .booking .left-panel {
  width: 70%;
  align-self: stretch;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
}
.c-flight-search-result .result .booking .left-panel .price {
  text-align: left;
  font-size: 24px;
  font-weight: bold;
  margin-bottom: 5px;
}
.c-flight-search-result .result .booking .left-panel .return-trip{
  display: flex;
  text-align: left;
  font-size: 16px;
  justify-content: space-between;
}
.c-flight-search-result .result .booking .left-panel .return-trip .going-trip {
  margin-right: 20px;
}
.c-flight-search-result .result .booking .left-panel .return-trip p{
  margin:0;
  margin-bottom: 4px;
}
.c-flight-search-result .result .booking .left-panel .return-trip p:first-child{
  font-size: 12px;
}
.c-flight-search-result .result .booking .right-panel {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-self: center;
}
.c-flight-search-result .result .booking .right-panel button {
  border-radius: 5px;
  border: 0;
  padding: 10px 20px;
  font-weight: bold;
  background-color: #ffad00;
  color: #000036;
  transition: all .5s;
  cursor: pointer;
  margin-top: 12px;
}
.c-flight-search-result .result .booking .right-panel button:hover {
  box-shadow: 1px 1px 10px rgba(0,0,0,.2);
}
.react-datepicker-wrapper {
  display: block;
  width: 100%;
}
.react-datepicker-wrapper .react-datepicker__input-container {
  display: block;
  width: 100%;
}
.react-datepicker-wrapper .react-datepicker__input-container input {
  width: 100%;
  box-sizing: border-box;
}
.react-datepicker__close-icon {
  display: flex;
}

@media (max-width: 767px){
  div#root {
    background-size: contain;
    background-position: top center;
    display: block;
  }
  main {
    width: 100%;
  }
  .c-search-engine {
    flex-direction: column;
    margin: 0 10px;
    margin-top: 68px;
    margin-bottom: 20px;
    padding-top: 30px;
  }
  .c-search-engine .c-flight-search {
    width: 100%;
    padding: 0;
  }
  .c-flight-search-result {
    width: 100%;
    background-color: transparent;
  }
  .c-flight-search-result .trip{
    padding-left: 0;
    color: #333;
  }
  .c-flight-search-result .result{
    padding:0;
  }
  .c-flight-search-result p.no-flight {
    color: #333;
  }
}
@media (min-width:768px) and (max-width: 1023px) {
  div#root {
    background-size: contain;
    background-position: top center;
  }
  .c-search-engine .c-flight-search {
    padding-right: 0;
    width: 38%;
  }
  .c-flight-search-result .result {
    padding-bottom: 20px;
  }
  .c-flight-search-result .result .booking .left-panel .return-trip {
    font-size: 14px;
  }
}