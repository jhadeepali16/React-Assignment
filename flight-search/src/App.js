import React, { Component } from 'react';
import FlightSearch from './component/FlightSearch';
import FlightSearchResult from'./component/FlightSearchResult';
import InputRange from 'react-input-range';
import PropTypes from "prop-types";
import 'react-input-range/lib/css/index.css';
import './App.css';

const url = 'http://localhost:3000/data';

class App extends Component {
  constructor(props){
    super(props);
    this.state = {
        requestFailed: false,
        isLoading: false,
        initialState: false,
        isRetrunTrip: false,
        isFocused: false,
        disabled: true,
        showError: false,
        noResult: false,
        data: '',
        filterData: '',
        pricefilterData: '',
        origin: '',
        destination: '',
        passenger:'',
        departureDate: null,
        returnDate: null,
        fieldActiveClass: '',
        toggleClass: 'inactive',
        activeElement: 'one-way',
        fieldId:{},
        error:{},
        maxPrice: 0,
        minPrice: 0,
        value: { min: '', max: '' },
        errMsg: 'Please fill the required field'
    }
  }
  getData = (filterItem) => {
    fetch(url)
    .then(response => {
        if (!response.ok) {
            throw Error("Network request failed")
        }
        return response
    })
    .then(d => d.json())
    .then(d => {
        this.setState({
            data: d,
            initialState: true,
            showError: false
        })
        this.searchData(filterItem);
    }, () => {
        this.setState({
            requestFailed: true
        })
    });
}
searchData = (filterItem) => {
    let result = [], ticketPrice, maxPrice, minPrice;
    if(this.state.initialState) {
        result = this.state.data;
        let show = result.filter(function(item){
            for(var key in filterItem){
                if(item[key] === undefined || item[key] !== filterItem[key])
                    return false;
            }
            return true;
        });
        if(show.length){
            ticketPrice = show.map(p => p.price);
            maxPrice = Math.max(...ticketPrice);
            minPrice = Math.min(...ticketPrice);
            this.setState({
                noResult: false,
                filterData: show,
                pricefilterData: show,
                maxPrice: maxPrice,
                minPrice: minPrice
            })
        }
        else this.setState({ noResult: true })
    }
}

isEmpty = (error) => {
    for(var key in error){
        if(error.hasOwnProperty(key))
            return false;
    }    
    return true;
}

validateField = () => {
    if(!this.state.disabled) {
        if(!this.state.origin && !this.state.destination && !this.state.returnDate && !this.state.departureDate && !this.state.passenger){
            this.setState({
                error : { ...this.state.error, origin: true, destination: true, returnDate: true, departureDate: true, passenger: true },
                errMsg: 'Please fill the required field'
            })
        }
        else {
            this.setState({
                error : { ...this.state.error, origin: false, destination: false, returnDate: false, departureDate: false, passenger: false },
                errMsg: ''
            })
        }
    }
    else {
        if(!this.state.origin && !this.state.destination && !this.state.departureDate && !this.state.passenger){
            this.setState({
                error : { ...this.state.error, origin: true, destination: true, departureDate: true, passenger: true },
                errMsg: 'Please fill the required field'
            })
        }
        else {
            this.setState({
                error : { ...this.state.error, origin: false, destination: false, departureDate: false, passenger: false },
                errMsg: ''
            })
        }
    }
}


handleSubmit = (event, key) => {
    let filterItem = {},
        error = {...this.state.error},
        keyValue = Object.values(error);
    
    this.validateField();
    event.preventDefault();
    if(Object.keys(error).length && keyValue.includes(false)) {
        filterItem["origin"] = this.state.origin;
        filterItem["destination"] = this.state.destination;
        this.getData(filterItem);
    }
    else this.setState({ showError: true })
}
handleChange = (event) => {
    const {name, value} = event.target;
    this.setState({
        [name]: value
    })
}
validateOnChange = (event) => {
    const {name, value} = event.target;
    if(value !== ""){
        this.setState(() => ({
            fieldId: {...this.state.fieldId, [name]: true },
            error: {...this.state.error, [name]: false }
        }))
    }
    else {
        this.setState(() => ({
            fieldId: {...this.state.fieldId, [name]: false },
            error: {...this.state.error, [name]: true }
        }))
    }
}

departureOnChange = (value) => {
    if(value !== ""){
        this.setState({
            departureDate: value,
            activeClass: 'active'
        })
    }
}
returnOnChange = (value) => {
    this.setState({
        returnDate: value,
        activeClass: 'active'
    })
}

changedateformat = (value) => {
    let d = new Date(value);
    return [d.getDate(), d.getMonth()+1, d.getFullYear()].join('-');
}

selectTrip = (e, index) => {
    let target = e.target.id;
    if(target === 'return-way'){
        this.setState({
            isRetrunTrip: true,
            disabled: false,
            toggleClass: 'active'
        });
    }
    else if (target === 'one-way'){
        this.setState({
            isRetrunTrip: false,
            disabled: true,
            toggleClass: 'inactive'
        })
    }
    this.setState({
        activeElement: target,
        error: {},
        fieldId: {},
        origin:'',
        destination:'',
        departureDate:'',
        returnDate:'',
        passenger:0,
        data: '',
        initialState: '',
        pricefilterData: '',
        noResult: ''
    })
}

priceSlider = value => this.setState({ value });

refinePrice = (value, min, max) => {
    max = value.max;
    min = value.min;
    let filterData = this.state.pricefilterData;
    let show = filterData.filter(function(item){
        return item.price >= min && item.price <= max; 
    });
    this.setState({ 
        filterData: show,
    });
}

render() {
    return (
      <main className="App">
        <div className="heading">
            <h1>Flight Search Engine</h1>
        </div>
        <section className="c-search-engine">
            <section className="c-flight-search">
                <FlightSearch validateOnChange={this.validateOnChange} handleSubmit={this.handleSubmit} handleChange={this.handleChange} selectTrip={this.selectTrip} onClickOutSide={this.onClickOutSide} isRetrunTrip={this.state.isRetrunTrip} activeClass={this.state.activeClass} activeTab={this.state.activeElement} fieldActiveClass={this.state.fieldActiveClass} toggleClass={this.state.toggleClass} origin={this.state.origin} destination={this.state.destination} passenger={this.state.passenger} departureOnChange={this.departureOnChange} returnOnChange={this.returnOnChange} fieldId={this.state.fieldId} error={this.state.error} departureDate={this.state.departureDate} returnDate={this.state.returnDate} isFocused={this.state.isFocused} maxPrice={this.state.maxPrice} minPrice={this.state.minPrice} value={this.state.value} handleOnBlur={this.handleOnBlur} disabled={this.state.disabled} errMsg={this.state.errMsg} />
                {this.state.initialState && this.state.filterData ? 
                    <section className="price-slider">
                        <h2>Refine Flight Search</h2>
                        <InputRange maxValue={parseInt(this.state.maxPrice, 10)} minValue={parseInt(this.state.minPrice, 10)} value={this.state.value} onChange={this.priceSlider} onChangeComplete={this.refinePrice} />
                    </section> : null
                }
            </section>
          <FlightSearchResult showError={this.state.showError} data={this.state.filterData} initialState={this.state.initialState} origin={this.state.origin} destination={this.state.destination} departureDate={this.state.departureDate} returnDate={this.state.returnDate} isRetrunTrip={this.state.isRetrunTrip} noResult={this.state.noResult} requestFailed={this.state.requestFailed} />
        </section>
      </main>
    );
  }
}
App.propTypes = {
    maxValue: PropTypes.number,
    minValue: PropTypes.number,
    value: PropTypes.object
}
export default App;
