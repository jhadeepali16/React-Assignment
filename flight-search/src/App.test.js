import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';
jest.mock('react-dom');

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders a snapshot', () => {
  const tree = renderer.create(<App/>).toJSON();
  expect(tree).toMatchSnapshot();
});
