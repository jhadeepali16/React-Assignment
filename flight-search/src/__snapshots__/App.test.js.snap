// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot 1`] = `
<main
  className="App"
>
  <div
    className="heading"
  >
    <h1>
      Flight Search Engine
    </h1>
  </div>
  <section
    className="c-search-engine"
  >
    <section
      className="c-flight-search"
    >
      <section>
        <section>
          <div
            className="element"
          >
            <div
              className="tabs active"
              id="one-way"
              onClick={[Function]}
            >
              One way
            </div>
            <div
              className="tabs "
              id="return-way"
              onClick={[Function]}
            >
              Return way
            </div>
          </div>
          <section
            className="form"
          >
            <div
              className="element"
            >
              <input
                className={null}
                id="origin"
                name="origin"
                onBlur={[Function]}
                onChange={[Function]}
                required="true"
                type="text"
                value=""
              />
              <label
                className={null}
                htmlFor="origin"
              >
                Enter Origin City
              </label>
              <span
                className="error"
              />
            </div>
            <div
              className="element"
            >
              <input
                className={null}
                id="destination"
                name="destination"
                onBlur={[Function]}
                onChange={[Function]}
                required="true"
                type="text"
                value=""
              />
              <label
                className={null}
                htmlFor="destination"
              >
                Enter Destination City
              </label>
              <span
                className="error"
              />
            </div>
            <div>
              <div
                className="react-datepicker-wrapper"
              >
                <div
                  className="react-datepicker__input-container"
                >
                  <div
                    className="element undefined"
                  >
                    <input
                      className={null}
                      disabled=""
                      id="departureDate"
                      name="departureDate"
                      onBlur={[Function]}
                      onChange={undefined}
                      onFocus={[Function]}
                      type={undefined}
                      value=""
                    />
                    <label
                      className={null}
                      htmlFor="departureDate"
                    >
                      Departure Date
                    </label>
                    <span
                      className="error"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="react-datepicker-wrapper"
              >
                <div
                  className="react-datepicker__input-container"
                >
                  <div
                    className="element inactive"
                  >
                    <input
                      className={null}
                      disabled="disabled"
                      id="returnDate"
                      name="returnDate"
                      onBlur={[Function]}
                      onChange={undefined}
                      onFocus={[Function]}
                      type={undefined}
                      value=""
                    />
                    <label
                      className={null}
                      htmlFor="returnDate"
                    >
                      Return Date
                    </label>
                    <span
                      className="error"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              className="element"
            >
              <select
                className={null}
                id="passenger"
                name="passenger"
                onBlur={[Function]}
                onChange={[Function]}
                required="true"
                value=""
              >
                <option
                  value=""
                >
                  
                </option>
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  value="3"
                >
                  3
                </option>
                <option
                  value="4"
                >
                  4
                </option>
                <option
                  value="5"
                >
                  5
                </option>
              </select>
              <label
                className={null}
                htmlFor="passenger"
              >
                Passengers
              </label>
              <span
                className="error"
              />
            </div>
            <div
              className="element"
            >
              <button
                className="btn-search"
                id="search"
                onClick={[Function]}
                type="submit"
              >
                Search
              </button>
            </div>
          </section>
        </section>
      </section>
    </section>
  </section>
</main>
`;
